### PR TITLE
Ticket11

### PR DESCRIPTION
**Pull request by yos244**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -187,17 +187,42 @@ describe("GET USERS", () => {
...
+      .get(`/api/reviews/3`)
+      .expect(200)
+      .then((commentObj) => {
+        expect(commentObj.body.comment_count).toBe(`3`);
+      });
   });
 });
 
-
+describe("GET Query", () => {
```
Comment (cjplatten) [inline __tests__/app.test.js:201]: There aren't any tests here for filtering by category, please add tests for that!

File: controller/controller.js [hunk 2]
```diff
@@ -18,8 +18,10 @@ exports.getCategories = (req, res, next) => {
 };
 
 exports.getReviews = (req, res, next) => {
-  selectReviews().then((reviews) => {
+  selectReviews(req.query).then((reviews) => {
```
Comment (cjplatten) [inline controller/controller.js:21]: I would recommend destructuring your queries here and passing them into the model individually, it will make them easier to interact with in the model.

File: model/model.js [hunk 3]
```diff
@@ -13,16 +14,56 @@ exports.selectCategories = () => {
...
+  if (query.hasOwnProperty(`category`)) {
+    return db
+      .query(
+        `
+      SELECT reviews.owner, reviews.title, reviews.review_id, reviews.category, reviews.review_img_url, reviews.created_at, reviews.votes, reviews.designer,  COUNT(comment_id) AS comment_count
+      FROM reviews
+      LEFT JOIN comments ON comments.review_id = reviews.review_id
+      WHERE category = $1
+      GROUP BY reviews.review_id
+      ORDER BY reviews.${sort_by} ${order_by};
```
Comment (cjplatten) [inline model/model.js]: Having the entire query string written out twice like this doesn't make for very DRY code, how might you build up the query string dynamically so that you only have to write one query?

File: model/model.js [hunk 4]
```diff
@@ -13,16 +14,56 @@ exports.selectCategories = () => {
...
+    if (
+      query.sort_by !== `title` &&
+      query.sort_by !== `designer` && 
+      query.sort_by !== `owner` &&
+      query.sort_by !== `review_img_url` &&
+      query.sort_by !== `review_body` &&
+      query.sort_by !== `category` &&
+      query.sort_by !== `created_at` &&
+      query.sort_by !== `votes`
+    ) {
```
Comment (cjplatten) [inline model/model.js:32]: This works fine but have a think about if there is a cleaner way to do this check with an array and an array method?

**Review by cjplatten — COMMENTED** [review]
Please see comments

File: __tests__/app.test.js [hunk 5]
```diff
@@ -189,6 +187,52 @@ describe("GET USERS", () => {
...
+      .then((sortedRev) => {
+        expect(sortedRev.body).toBeSortedBy(`votes`, { descending: true });
+      });
+  });
+  test("GET: 200 order_by and sort_by", () => {
+    return request(app)
+      .get(`/api/reviews?sort_by=votes&order_by=asc`)
+      .expect(200)
+      .then((sortedRev) => {
+        expect(sortedRev.body).toBeSortedBy(`votes`);
```
Comment (Geo-Beard) [inline __tests__/app.test.js:226]: Not sure if you're missing {ascending: true} here?

File: __tests__/app.test.js [hunk 6]
```diff
@@ -189,6 +187,52 @@ describe("GET USERS", () => {
...
+  });
+});
+
+describe("GET Query", () => {
+  test("GET: 200 Query with category", () => {
+    return request(app)
+      .get(`/api/reviews?category=social+deduction`)
+      .expect(200)
+      .then((revCat) => {
+        expect(revCat.body.length).toBe(11);
```
Comment (Geo-Beard) [inline __tests__/app.test.js:207]: Need to check that all returned objects have the correct category here - not just the length.

File: __tests__/app.test.js [hunk 7]
```diff
@@ -189,6 +187,52 @@ describe("GET USERS", () => {
...
+      .get(`/api/reviews/3`)
+      .expect(200)
+      .then((commentObj) => {
+        expect(commentObj.body.comment_count).toBe(`3`);
+      });
+  });
+});
+
+describe("GET Query", () => {
+  test("GET: 200 Query with category", () => {
```
Comment (Geo-Beard) [inline __tests__/app.test.js:202]: Also need a status 200 test that takes a valid category that has no reviews - should return an empty array.

File: controller/controller.js [hunk 8]
```diff
@@ -68,3 +70,12 @@ exports.getUsers = (req, res, next) => {
     res.status(200).send(users);
   });
 };
+
+
+exports.deleteComment = (req,res,next) => {
+  removeComment(req.params.comment_id).then((deletedComment)=> {
+    res.status(204).send(deletedComment)
```
Comment (Geo-Beard) [inline controller/controller.js]: Just respond with the 204 - and don't send anything back at all. Change to .send().

File: __tests__/app.test.js [hunk 9]
```diff
@@ -189,6 +187,52 @@ describe("GET USERS", () => {
...
+  });
+});
+
+describe("DELETE api", () => {
+  test("DELETE 204 - Delete comment given comment ID", () => {
+    return request(app)
+      .delete(`/api/comments/4`)
+      .expect(204)
+      .then((deletedComment) => {
+        expect(deletedComment.body).toEqual({});
```
Comment (Geo-Beard) [inline __tests__/app.test.js]: See my notes in the controller - you shouldn't be sending anything back so you shouldn't test that an empty object is returned.

File: testdb.sql [hunk 10]
```diff
@@ -0,0 +1,23 @@
+
+\c nc_games_test
```
Comment (Geo-Beard) [inline testdb.sql]: Remove commented out code.

**Review by Geo-Beard — COMMENTED** [review]
I've reviewed changes you made that implemented Carrie's comments for ticket 11 and looked at ticket 12 (delete) as well. Ideally these should be separate tickets.

For ticket 11 changes - you need to look at my comments on your tests and implement those changes.

For ticket 12 - there are a few things you need to change and look at as well. See comments

When you've implemented these changes submit another PR (preferably separately if possible) for ticket 11 and 12 :)

File: __tests__/app.test.js (hunk 5 above)
Comment (yos244) [reply __tests__/app.test.js:226]: nope it is asc by default

File: __tests__/app.test.js (hunk 6 above)
Comment (Geo-Beard) [reply __tests__/app.test.js:207]: This is the social+deduction I was talking about - should be social%20deduction I believe
